storage/range.go

```diff
@@ -278,20 +279,76 @@ func (r *Range) IsFirstRange() bool {
 	return bytes.Equal(r.Desc().StartKey, engine.KeyMin)
 }
 
-// IsLeader returns true if this range replica is the raft leader.
-// TODO(spencer): this is always true for now.
-func (r *Range) IsLeader() bool {
-	return true
+// getLease returns the current leader lease.
+func (r *Range) getLease() *proto.Lease {
+	return (*proto.Lease)(atomic.LoadPointer(&r.lease))
+}
+
+// HasLeaderLease returns true if this range replica currently holds
+// the Raft leader lease. It's possible that the replica is not in
+// fact the leader, but holding the lease means that no other replica
```

Comment:
The instantaneous check for writes is enough, by my reasoning. It'll gate any reads that come after it on the node, which is all it needs to do. If the write completes, that means we still were leader, which is just fine. If we lost lease during write, either reads succeeded or they didn't while lease was definitively held (see next paragraph for read behavior). If the write fails from loss of leadership, then no harm done. Certainly no need to check the lease _after_ write has succeeded.

For reads, we'll need to return a NotLeaderError if we discover that we've lost the lease while reading the data. If we don't do this, then there's a chance that another leader has been elected and has accepted writes at the key we just read (making our read inconsistent). Will update this CL to include that check. It's really just moving the current check we do in addReadOnlyCmd until after we've done the read.